[PrintAsObjC] Use 'unsafe_unretained' to print 'unowned', not 'assign'

jrose-apple opened the pull request:
The ObjC generator previously preserved a subtle difference between `unowned` and `unowned(unsafe)` / Unmanaged by printing the former as `assign` and the latter as `unsafe_unretained`. Upstream Clang, however, has gotten a new warning to discourage the use of `assign` with reference-countable types at all. Since it was always a subtle distinction, just go with the new convention and print `unsafe_unretained` for `unowned` properties as well.

rdar://problem/44290715

jrose-apple committed:
[PrintAsObjC] Use 'unsafe_unretained' to print 'unowned', not 'assign'

The ObjC generator previously preserved a subtle difference between
'unowned' and 'unowned(unsafe)' / Unmanaged by printing the former as
'assign' and the latter as 'unsafe_unretained'. Upstream Clang,
however, has gotten a new warning to discourage the use of 'assign'
with reference-countable types at all. Since it was always a subtle
distinction, just go with the new convention and print
'unsafe_unretained' for 'unowned' properties as well.

rdar://problem/44290715

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
@swift-ci Please test macOS

rjmccall commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1077,9 +1074,11 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+      // We treat "unowned" as "unsafe_unretained" (even though it's more like
+      // "safe_unretained") because we want people to think twice about
+      // allowing that object to disappear. "unowned(unsafe)" (and Unmanaged,
+      // handled below) really are "unsafe_unretained".
```
Can this reasonably be turned into an exhaustive switch over reference storage kinds?

rjmccall reviewed (approved):
Comment aside, LGTM.  I filed rdar://44290715 before I caught up on my code-review inbox and saw that you'd done work on this, though, sorry.

jrose-apple committed:
[PrintAsObjC] Refactor printing of property ownership attributes

Exhaustively switch over reference types, so that we'll know if any
new ones are added, and factor out a repeated check to see if
something's a retainable object type.

No functionality change.

jrose-apple commented:
Good suggestion, spotted another reasonable cleanup while I was doing that.

@swift-ci Please smoke test
